### ISSUE TITLE
Impossible to hook into suspend-then-hibernate

### ISSUE THREAD

**Issue by gentoo-root** [systemd/systemd#18092]
**systemd version the issue has been seen with**
 > 247

**Used distribution**
 > Gentoo

**Linux kernel version used** (`uname -a`)
 > Linux … 5.10.3-gentoo #1 SMP PREEMPT Sun Dec 27 21:32:36 EET 2020 x86_64 Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz GenuineIntel GNU/Linux

**CPU architecture issue was seen on**
 > x86_64 (any)

**Expected behaviour you didn't see**
As a user, I want to be able to write hooks (scripts) for pre-suspend and post-suspend events. This is currently possible for normal suspend, hibernate and hybrid sleep in two ways: either by putting them into /lib/systemd/system-sleep/, or by writing service unit files. In particular, it's useful for doing hardware-specific actions before entering and after leaving the low power modes (for example, [nvidia power management](https://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/powermanagement.html)). However, this functionality is not available for suspend-then-hibernate: it's not possible to create "before suspend", "after resume", "before hibernate" and "after thaw" hooks for this mode.

**Unexpected behaviour you saw**
Custom unit files with `Before=systemd-suspend-then-hibernate.service` and `After=systemd-suspend-then-hibernate.service` run only before and after the whole process, so they are not suitable to run hardware-specific actions, because it's not clear whether the system is resuming or thawing, and there is no place to hook into the transition from suspend to hibernate. The scripts in the /lib/systemd/system-sleep/ directory also don't fit, because they get arguments of `pre/post suspend-then-hibernate`, and it's not possible to distinguish between suspend and hibernate.

**Steps to reproduce the problem**
1. Create a script and put it into /lib/systemd/system-sleep/test (make it executable):
```
#!/bin/sh

logger -t 'sleep-debug' "$1" "$2"
```
2. Create a custom unit and put it into /etc/systemd/system/sleep-debug-before.service:
```
[Unit]
Description=sleep debug before
Before=systemd-suspend-then-hibernate.service

[Service]
Type=oneshot
ExecStart=/usr/bin/logger -t sleep-debug before

[Install]
RequiredBy=systemd-suspend-then-hibernate.service
```
3. Create a custom unit and put it into /etc/systemd/system/sleep-debug-after.service:
```
[Unit]
Description=sleep debug after
After=systemd-suspend-then-hibernate.service

[Service]
Type=oneshot
ExecStart=/usr/bin/logger -t sleep-debug after

[Install]
RequiredBy=systemd-suspend-then-hibernate.service
```
4. `systemctl suspend-then-hibernate`
5. Wait for HibernateDelaySec and let the system hibernate.
6. Turn on the computer and let it thaw.
7. `journalctl -ab -t sleep-debug`

You'll see that there were two "pre suspend-then-hibernate" and two "post suspend-then-hibernate" messages, and it's not possible to distinguish which of them are for suspend and which are for hibernate. Moreover, if the hibernate stage fails, systemd-sleep falls back to suspending again, which only complicates the situation. Also you'll see that "before" and "after" messages are printed before and after the whole process (before suspend and after hibernate), so these services are also not suitable to do hardware-specific preparations.

**Comment by zachsmith** [comment]
@gentoo-root would you mind testing the proposed fix found in https://github.com/systemd/systemd/pull/18110?

**Comment by gentoo-root** [comment]
@zachsmith, sure, I'll test it and get back to you with the results, thanks a lot for the prompt fix!

While your fix should be fine for my purpose, I have one more thought, though. systemd-sleep(8) mentions the following:

>Note that scripts or binaries dropped in /usr/lib/systemd/system-sleep/ are intended for local use only and should be considered hacks.

And unlike other config dirs in systemd, this one doesn't have a location in /etc for user-specific configs and possible overrides of configs from /usr/lib. Also, I'm not sure if running hardware-specific scripts on suspend/resume is considered a hack, or should it use another mechanism like unit files. It's hard for me to tell, because I don't know why exactly this systemd-sleep directory was added. This whole mechanism of suspend in systemd might need some rethinking, some ideas are:

1. Extend the /usr/lib/systemd/system-sleep/ interface to add a directory under /etc. Maybe integrate it better with units somehow.
2. Replace suspend-then-hibernate.target with three different targets representing different steps of suspend-then-hibernate, then the user can create units with the right dependencies and integrate well with the system of units (like having all the functionality such as conditions, dropping privileges, sandboxing, etc.). IMO, it should fit better into the philosophy of systemd than bash scripts in a single directory.

Please tell me what you think of it.

**Comment by bluca** [comment]
> This whole mechanism of suspend in systemd might need some rethinking, some ideas are:
> 
> 1. Extend the /usr/lib/systemd/system-sleep/ interface to add a directory under /etc. Maybe integrate it better with units somehow.
> 2. Replace suspend-then-hibernate.target with three different targets representing different steps of suspend-then-hibernate, then the user can create units with the right dependencies and integrate well with the system of units (like having all the functionality such as conditions, dropping privileges, sandboxing, etc.). IMO, it should fit better into the philosophy of systemd than bash scripts in a single directory.
> 
> Please tell me what you think of it.

I'd suggest to open a separate issue for this as an RFE

**Comment by poettering** [comment]
> Also, I'm not sure if running hardware-specific scripts on suspend/resume is considered a hack,

It is. Correctly, cleanly written drivers do not require that.

**Comment by poettering** [comment]
Let's not extend this iface. it's supposed to be an interface for doing quick hacks, workar-ounds for shitty drivers. it's not supposed to be a first class interface.

it's fine to pass more metadata along via env var, as discussed in that other PR, but I am strongly against making this a first class citizen

**Comment by jes-n** [comment]
I want to share my hack for solving this issue, which involves a script in /usr/lib/systemd/system-sleep. Unfortunately, I'm using systemd-v245 in Ubuntu 20.04 and not able to patch systemd with the fix provided by @zachsmith in #18110 with the `SYSTEMD_SLEEP_ACTION` environment variable.

This hack works by writing the expected time the system will wake itself to /root/.suspend-then-hibernate when executing the "pre" hook. This expected time is determined using the "HibernateDelaySec=" option in /etc/systemd/sleep.conf along with the current time given by the `date` utility. When the script is again executed with the "post" hook, this /root/.suspend-then-hibernate file is checked. If it exists _and_ the current time given by `date` minus the time written to /root/.suspend-then-hibernate is within the given `window`, it can be assumed the system is waking itself up from suspend to hibernate. Otherwise, it is assumed the system was woken up from suspend prior to HibernateDelaySec or from hibernate.

```bash
#!/bin/bash

# Define a file to store the expected time the system will wake-up from suspend to hibernate (can be any persistent file).
lock=/root/.suspend-then-hibernate

# Determine the expected wake-up time by extracting 'HibernateDelaySec=' from /etc/systemd/sleep.conf
# NOTE: If 'HibernateDelaySec=' is not matched in /etc/systemd/sleep.conf, the 2-hour default is assumed.
#   https://www.freedesktop.org/software/systemd/man/sleep.conf.d.html#HibernateDelaySec=
delay=$(grep "^HibernateDelaySec=" /etc/systemd/sleep.conf || echo "HibernateDelaySec=2hour")

# Define a "time-window" in which the *post* action is assumed to be the system waking itself up to hibernate,
# as opposed to being woken-up from suspend prior to HibernateDelaySec.
# The value of this window is system-depenedent, relating to how quickly the system wakes itself up from suspend and
# executes hibernate. In testing on my own modern system (Ryzen 7, 20GB RAM, NVMe boot drive), this was consistently
# between 2 and 5 seconds.
window=10

case $1/$2 in
  pre/suspend-then-hibernate)
    # Store the expected time the system will wake-up to hibernate, determined using the `date` utility
    echo $(date -d "+${delay#HibernateDelaySec=}" +%s) > $lock

    # >>> Perform pre-suspend/pre-hibernate actions (not possible to tell which without the fix in systemd issue #18110)
    echo "Entering suspend/hibernate"
    ;;

  post/suspend-then-hibernate)
    # If $lock exists *and* the current time is within $window, assume the system is waking itself up to hibernate
    if [ -f  "$lock" ] && [ $((`date +%s` - `cat $lock`)) -lt $window ]; then
      # >>> Perform post-suspend/pre-hiberate actions
      echo "Waking-up from suspend to hibernate"

    # Otherwise, assume the system is being woken-up from suspend prior to HibernateDelaySec
    else
      # >>> Perform post-suspend/post-hibernate actions
      echo "Waking-up from suspend/hibernate to power-on"

    fi
    ;;
esac
```

The following assumptions must be made:
1. If "HibernateDelaySec=" is specified in /etc/systemd/sleep.conf, it is written in a format parsable by the `date` utility using the `-d|--date` option (for example, '1hour' works but '1h' does not).
2. The system using this script will wake itself up from suspend and enter hibernate within the number of seconds set by the `window` variable.
3. The script is executed using Bash (for variable substring removal and arithmetic capability) and has access to the `date`, `grep`, and `cat` utilities.

I use this script for power savings in a Slurm cluster, where nodes are initially powered down with suspend before going into hibernate after 8 hours. When the nodes kick back in to enter hibernate, the Slurm controller node receives an unexpected signal from them during the "post" suspend, "pre" hibernate phase. This causes the nodes to be marked as DOWN, preventing them from being correctly woken up again.

This hack is to help anyone using a version of systemd prior to the fix in #18110.
